commands/audit/sca/pnpm/pnpm.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// Install is required when "pnpm-lock.yaml" lock file or "node_modules/.pnpm" directory not exists.
+func installProjectIfNeeded(pnpmExecPath, workingDir string) (err error) {
+	lockFileExists, err := fileutils.IsFileExists(filepath.Join(workingDir, "pnpm-lock.yaml"), false)
+	if err != nil {
+		return
+	}
+	pnpmDirExists, err := fileutils.IsDirExists(filepath.Join(workingDir, "node_modules", ".pnpm"), false)
+	if err != nil || (lockFileExists && pnpmDirExists) {
+		return
+	}
+	// Install is needed
+	log.Debug("Installing Pnpm project:", workingDir)
+	return getPnpmCmd(pnpmExecPath, workingDir, "install").GetCmd().Run()
+}
+
+// Run 'pnpm ls ...' command (project must be installed) and parse the returned result to create a dependencies map of.
+func calculateDependencies(executablePath, workingDir string, params utils.AuditParams) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string, err error) {
+	lsArgs := append([]string{"--depth", "Infinity", "--json", "--long"}, params.Args()...)
+	npmLsCmdContent, err := getPnpmCmd(executablePath, workingDir, "ls", lsArgs...).RunWithOutput()
+	if err != nil {
+		return
+	}
+	log.Debug("Pnpm ls command output:\n", string(npmLsCmdContent))
+	output := &[]pnpmLsProject{}
+	if err = json.Unmarshal(npmLsCmdContent, output); err != nil {
+		return
+	}
+	dependencyTrees, uniqueDeps = parsePnpmLSContent(*output)
+	return
+}
+
+func parsePnpmLSContent(projectInfo []pnpmLsProject) (dependencyTrees []*xrayUtils.GraphNode, uniqueDeps []string) {
+	uniqueDepsSet := datastructures.MakeSet[string]()
+	for _, project := range projectInfo {
+		// Parse the dependencies into Xray dependency tree format
+		dependencyTree, uniqueProjectDeps := coreXray.BuildXrayDependencyTree(createProjectDependenciesTree(project), getDependencyId(project.Name, project.Version))
+		// Add results
+		dependencyTrees = append(dependencyTrees, dependencyTree)
+		uniqueDepsSet.AddElements(maps.Keys(uniqueProjectDeps)...)
+	}
+	uniqueDeps = uniqueDepsSet.ToSlice()
+	return
+}
+
+func createProjectDependenciesTree(project pnpmLsProject) map[string]coreXray.DepTreeNode {
+	treeMap := make(map[string]coreXray.DepTreeNode)
+	directDependencies := []string{}
+	// Handle production-dependencies
+	for depName, dependency := range project.Dependencies {
+		directDependency := getDependencyId(depName, dependency.Version)
+		directDependencies = append(directDependencies, directDependency)
+		appendTransitiveDependencies(directDependency, dependency.Dependencies, treeMap)
+	}
+	// Handle dev-dependencies
+	for depName, dependency := range project.DevDependencies {
+		directDependency := getDependencyId(depName, dependency.Version)
+		directDependencies = append(directDependencies, directDependency)
+		appendTransitiveDependencies(directDependency, dependency.Dependencies, treeMap)
+	}
+	if len(directDependencies) > 0 {
+		treeMap[getDependencyId(project.Name, project.Version)] = coreXray.DepTreeNode{Children: directDependencies}
```

Comment:
I'm just curious - What is the reason that you initialize the treeMap only if there are dependencies?

Reply:
It's the same for other package-managers.
If there are no dependencies in the project they return an empty map. not a map with only one empty entry for the project id.